exporter/trace/zipkin/zipkin.go

```diff
@@ -0,0 +1,194 @@
+	return &Exporter{
+		reporter:      http.NewReporter(zipkinURL, opts...),
+		localEndpoint: localEndpoint,
+	}, nil
+}
+
+// Export exports a span to a Zipkin server.
+func (e *Exporter) Export(s *trace.SpanData) {
+	e.reporter.Send(zipkinSpan(s, e.localEndpoint))
+}
+
+const (
+	statusCodeTagKey        = "census.status_code"
+	statusDescriptionTagKey = "census.status_description"
+)
+
+var sampledTrue = true
+
+func convertTraceID(t trace.TraceID) model.TraceID {
+	return model.TraceID{
+		High: binary.BigEndian.Uint64(t[:8]),
+		Low:  binary.BigEndian.Uint64(t[8:]),
+	}
+}
+
+func convertSpanID(s trace.SpanID) model.ID {
+	return model.ID(binary.BigEndian.Uint64(s[:]))
+}
+
+func spanKind(s *trace.SpanData) model.Kind {
+	if s.HasRemoteParent {
+		return model.Server
+	}
+	if strings.HasPrefix(s.Name, "Sent.") {
+		return model.Client
+	}
+	if strings.HasPrefix(s.Name, "Recv.") {
+		return model.Server
+	}
+	if len(s.MessageEvents) > 0 {
+		switch s.MessageEvents[0].EventType {
+		case trace.MessageEventTypeSent:
+			return model.Client
+		case trace.MessageEventTypeRecv:
+			return model.Server
+		}
+	}
+	return model.Undetermined
+}
+
+func zipkinSpan(s *trace.SpanData, localEndpoint *model.Endpoint) model.SpanModel {
+	sc := s.SpanContext
+	z := model.SpanModel{
+		SpanContext: model.SpanContext{
+			TraceID: convertTraceID(sc.TraceID),
+			ID:      convertSpanID(sc.SpanID),
+			Sampled: &sampledTrue,
+		},
+		Kind:          spanKind(s),
+		Name:          s.Name,
+		Timestamp:     s.StartTime,
+		Shared:        false,
+		LocalEndpoint: localEndpoint,
+	}
+
+	if s.ParentSpanID != (trace.SpanID{}) {
+		id := convertSpanID(s.ParentSpanID)
+		z.ParentID = &id
+	}
+
+	if s, e := s.StartTime, s.EndTime; !s.IsZero() && !e.IsZero() {
+		z.Duration = e.Sub(s)
+	}
+
+	// construct Tags from s.Attributes and s.Status.
+	if len(s.Attributes) != 0 {
+		m := make(map[string]string, len(s.Attributes)+2)
+		for key, value := range s.Attributes {
+			switch v := value.(type) {
+			case string:
+				m[key] = v
+			case bool:
+				if v {
+					m[key] = "true"
+				} else {
+					m[key] = "false"
+				}
+			case int64:
+				m[key] = strconv.FormatInt(v, 10)
+			}
+		}
+		z.Tags = m
+	}
+	if s.Status.Code != 0 || s.Status.Message != "" {
+		if z.Tags == nil {
+			z.Tags = make(map[string]string, 2)
+		}
+		z.Tags[statusCodeTagKey] = code(s.Status.Code).String()
+		if s.Status.Message != "" {
+			z.Tags[statusDescriptionTagKey] = s.Status.Message
+		}
+	}
+
+	// construct Annotations from s.Annotations and s.MessageEvents.
+	z.Annotations = make([]model.Annotation, 0, len(s.Annotations)+len(s.MessageEvents))
+	for _, a := range s.Annotations {
+		z.Annotations = append(z.Annotations, model.Annotation{
+			Timestamp: a.Time,
+			Value:     a.Message,
+		})
+	}
+	for _, m := range s.MessageEvents {
+		a := model.Annotation{
+			Timestamp: m.Time,
+		}
+		switch m.EventType {
+		case trace.MessageEventTypeSent:
```

Comment:
if sent/receive are about span.kind, I'd leave it out and just substitute span.timestamp/duration. If they are enclosed and not the same, then leave in.

Reply:
This is to create annotations for individual message events, if they're in the census span.